adapters/AppLovin/AppLovinAdapter/GADMAdapterAppLovin.m

```diff
@@ -0,0 +1,227 @@
+static bool loggingEnabled = NO;
+
++ (void)initialize
+{
+    ALGADAdsQueue = [ALGADQueue queue];
+}
+
+- (instancetype)initWithGADMAdNetworkConnector:(id<GADMAdNetworkConnector>)connector
+{
+    self = [super init];
+    if ( self )
+    {
+        self.connector = connector;
+        self.sdk = [GADMAdapterAppLovinUtils sdkForCredentials: _connector.credentials];
+        
+        if ( !self.sdk )
+        {
+            [self log: @"Failed to initialize SDK"];
+        }
+    }
+    return self;
+}
+
++ (NSString *)adapterVersion
+{
+    return kGADMAdapterAppLovinVersion;
+}
+
++ (Class<GADAdNetworkExtras>)networkExtrasClass
+{
+    return [GADMAdapterAppLovinExtras class];
+}
+
+#pragma mark Interstitial Methods
+
+- (void)getInterstitial
+{
+    self.placement = [GADMAdapterAppLovinUtils placementFromCredentials: self.connector.credentials];
+    
+    // If we already have preloaded ads, don't fire off redundant requests
+    if ( ALGADAdsQueue.count < ALGADAdsQueueMinCapacity )
+    {
+        [self log: @"Requesting AppLovin interstitial for placement: %@", self.placement];
+        [self.sdk.adService loadNextAd: [ALAdSize sizeInterstitial] andNotify: self];
+        //    [NSTimer scheduledTimerWithTimeInterval: 10.0 target: self selector: @selector(loadAfterSleep:) userInfo: nil repeats: NO];
+    }
+    else
+    {
+        [self log: @"Attempting to get another interstitial ad when we have preloaded ads already for placement: %@", self.placement];
+        
+        // Simulate ad load netowrk latency
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+            [self.connector adapterDidReceiveInterstitial: self];
+        });
+    }
+}
+
+- (void)loadAfterSleep:(NSTimer *)timer
+{
+    [self.sdk.adService loadNextAd: [ALAdSize sizeInterstitial] andNotify: self];
+}
+
+- (void)presentInterstitialFromRootViewController:(UIViewController *)rootViewController
+{
+    ALAd *dequeuedAd = [ALGADAdsQueue dequeue];
+    if ( dequeuedAd )
+    {
+        GADMAdapterAppLovinExtras *extras = _connector.networkExtras;
+        if ( extras && extras.muteAudio )
+        {
+            self.sdk.settings.muted = YES;
+        }
+        
+        [self log: @"Showing interstitial for placement: %@", self.placement];
+        [self.interstitial showOver: [UIApplication sharedApplication].keyWindow placement: self.placement andRender: dequeuedAd];
+    }
+    else
+    {
+        [self log: @"Failed to show an AppLovin interstitial before one was loaded"];
+        
+        NSError *error = [NSError errorWithDomain: kGADMAdapterAppLovinErrorDomain
+                                             code: kALErrorCodeUnableToRenderAd
+                                         userInfo: @{NSLocalizedFailureReasonErrorKey : @"Adaptor requested to display an interstitial before one was loaded"}];
+        
+        [self.connector adapter: self didFailAd: error];
```

Comment:
The `presentInterstitialFromRootViewController:` method is only called if an `adapterDidReceiveInterstitial:` callback has been sent. Ideally this block should never happen but  in a rare case that this happens, we need to send ad open and ad closed events, which a publisher might depend on to resume their app after requesting to show an ad. 

Please see the [Facebook Rewarded video adapter](https://github.com/googleads/googleads-mobile-ios-mediation/blob/master/adapters/Facebook/FacebookAdapter/GADFBRewardedVideoAd.m#L116) which uses a similar logic.